internal/driver/webui.go

```diff
@@ -0,0 +1,234 @@
+	"github.com/google/pprof/internal/plugin"
+	"github.com/google/pprof/internal/report"
+	"github.com/google/pprof/profile"
+)
+
+// webUI holds the state needed for serving a browser based interface.
+type webUI struct {
+	prof    *profile.Profile
+	options *plugin.Options
+}
+
+func serveWebInterface(port int, p *profile.Profile, o *plugin.Options) error {
+	interactiveMode = true
+	ui := &webUI{
+		prof:    p,
+		options: o,
+	}
+	// authorization wrapper
+	wrap := o.HTTPWrapper
+	if wrap == nil {
+		// only allow requests from local host
+		wrap = checkLocalHost
+	}
+	http.Handle("/", wrap(http.HandlerFunc(ui.dot)))
+	http.Handle("/disasm", wrap(http.HandlerFunc(ui.disasm)))
+	http.Handle("/weblist", wrap(http.HandlerFunc(ui.weblist)))
+	go openBrowser(port)
+	return http.ListenAndServe(fmt.Sprint(":", port), nil)
+}
+
+func openBrowser(port int) {
+	// Construct URL.
+	u, _ := url.Parse(fmt.Sprint("http://localhost:", port))
+	q := u.Query()
+	for _, p := range []struct{ param, key string }{
+		{"f", "focus"},
+		{"s", "show"},
+		{"i", "ignore"},
+		{"h", "hide"},
+	} {
+		if v := pprofVariables[p.key].value; v != "" {
+			q.Set(p.param, v)
+		}
+	}
+	u.RawQuery = q.Encode()
+
+	// Give server a little time to get ready.
+	time.Sleep(time.Millisecond * 500)
+
+	for _, b := range browsers() {
+		args := strings.Split(b, " ")
+		if len(args) == 0 {
+			continue
+		}
+		viewer := exec.Command(args[0], append(args[1:], u.String())...)
+		viewer.Stderr = os.Stderr
+		if err := viewer.Start(); err == nil {
+			return
+		}
+	}
+	// No visualizer succeeded, so just print URL.
+	fmt.Fprintln(os.Stderr, "Visit", u.String(), "in a web browser")
+}
+
+func checkLocalHost(h http.Handler) http.Handler {
+	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		host, _, err := net.SplitHostPort(req.RemoteAddr)
+		if err != nil || ((host != "127.0.0.1") && (host != "::1")) {
+			http.Error(w, "permission denied", http.StatusForbidden)
+			return
+		}
+		h.ServeHTTP(w, req)
+	})
+}
+
+// disasm generates a web page containing an svg diagram..
+func (ui *webUI) dot(w http.ResponseWriter, req *http.Request) {
+	if req.URL.Path != "/" {
+		http.NotFound(w, req)
+		return
+	}
+
+	// Generate dot graph.
+	args := []string{"svg"}
+	vars := pprofVariables.makeCopy()
+	vars["focus"].value = req.URL.Query().Get("f")
+	vars["show"].value = req.URL.Query().Get("s")
+	vars["ignore"].value = req.URL.Query().Get("i")
+	vars["hide"].value = req.URL.Query().Get("h")
+	_, rpt, err := generateRawReport(ui.prof, args, vars, ui.options)
+	if err != nil {
+		reportHandlerError(w, "could not generate report", err)
+		return
+	}
+	g, config := report.GetDOT(rpt)
+	legend := config.Labels
+	config.Labels = nil
+	dot := &bytes.Buffer{}
+	graph.ComposeDot(dot, g, &graph.DotAttributes{}, config)
+
+	// Convert to svg.
+	svg, err := dotToSvg(dot.Bytes())
+	if err != nil {
+		reportHandlerError(w, "Failed to execute dot. Is Graphviz installed?", err)
```

Comment:
Maybe change to "failed to execute dot: is Graphviz installed?" to start with lowercase like in other messages.